### PR TITLE
Recorder is fast now

### PR DESCRIPTION
**Pull request by hobinjk-ptc**
tl;dr (more in commit message): diffing is 16x faster

### CODE REVIEW

File: libraries/recorder.js [hunk 1]
```diff
@@ -168,20 +172,29 @@ recorder.update = function() {
...
+ * @param {number} targetTime
+ * @param {object|undefined} checkpoint
+ */
 recorder.replay = function(timeObject, targetTime, checkpoint) {
-    let objects = {};
-    let currentTime = 0;
-    if (checkpoint) {
+    let times = Object.keys(timeObject).map(t => parseInt(t));
+
+    let objects = JSON.parse(JSON.stringify(timeObject[times[0]]));
```
Comment (ptc-mklaudiny) [inline libraries/recorder.js:184]: Is stringify -> parse need for the undefined/null issue or is there other reason?

File: libraries/recorder.js [hunk 2]
```diff
@@ -208,62 +229,63 @@ recorder.applyDiff = function(objects, diff) {
 };
 
 /**
- * @param {object} obj - view into recorder.object
- * @param {object} objectInTime - object to store any values that change
+ * @param {object} cursorObj - view into recorder.object
+ * @param {object} cursorObj - view into recorder.objectOld
```
Comment (ptc-mklaudiny) [inline libraries/recorder.js:233]: How about cursorObjOld ?

File: libraries/recorder.js [hunk 3]
```diff
@@ -208,62 +229,63 @@ recorder.applyDiff = function(objects, diff) {
...
+            if (item !== cursorOld[key]) {
+                cursorTime[key] = item;
+                altered = true;
             }
 
-            oldItem[key] = thisValue;
+            cursorOld[key] = item;
         }
     }
+    return altered;
```
Comment (ptc-mklaudiny) [inline libraries/recorder.js:288]: recorder.getItemFromArray below can be removed

File: tests/recorder.test.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
+/* global test, expect */
+
+const recorder = require('../libraries/recorder.js');
```
Comment (ptc-mklaudiny) [inline tests/recorder.test.js:3]: So nice to see a unit test! :)

File: tests/recorder.test.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    objects.a[1] = 12;
+    recorder.saveState(1);
+
+    expectedTimeObject[1] = { a: [ undefined, 12 ] };
+    expect(recorder.timeObject).toEqual(expectedTimeObject);
+
+    // Delete b
+    delete objects.b;
+    recorder.saveState(2);
+    // No modification, deletions aren't reflected in diff
```
Comment (ptc-mklaudiny) [inline tests/recorder.test.js:61]: Now I realise that we do not get the exact state of var objects at a given time by replay mechanism. In practice, how do we then distinguish whether an object did not change for some time or whether it is already deleted?

**Review by ptc-mklaudiny — COMMENTED** [review]
I never quite understood the keyString 'path' addressing of items. This is clearer to me now.
16x speedup is great!

File: libraries/recorder.js (hunk 3 above)
Comment (hobinjk-ptc) [reply libraries/recorder.js:288]: Thanks!

File: tests/recorder.test.js (hunk 5 above)
Comment (hobinjk-ptc) [reply tests/recorder.test.js:61]: I believe most cases use an idle timeout type approach "if object not updated in X milliseconds then it's deleted" but this is definitely an area where I don't remember the exact particulars

File: libraries/recorder.js (hunk 1 above)
Comment (ptc-mklaudiny) [reply libraries/recorder.js:164]: ah, right.

**Review by ptc-mklaudiny — APPROVED** [review]

**Pull request merged**